module/zfs/mmp.c

```diff
@@ -0,0 +1,316 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright (c) 2017 by Lawrence Livermore National Security, LLC.
+ */
+
+#include <sys/abd.h>
+#include <sys/dsl_pool.h>
+#include <sys/mmp.h>
+#include <sys/spa.h>
+#include <sys/spa_impl.h>
+#include <sys/vdev.h>
+#include <sys/vdev_impl.h>
+#include <sys/zfs_context.h>
+
+static void mmp_thread(dsl_pool_t *dp);
+
+uint zfs_mmp_interval = 1000;		/* time between mmp writes in ms */
+uint zfs_mmp_fail_intervals = 5;	/* safety factor for MMP writes */
+
+void
+mmp_init(dsl_pool_t *dp)
+{
+	mmp_thread_state_t *mmp = &dp->dp_mmp;
+
+	bzero(mmp, sizeof (mmp_thread_state_t));
+	mutex_init(&mmp->mmp_thread_lock, NULL, MUTEX_DEFAULT, NULL);
+	cv_init(&mmp->mmp_thread_cv, NULL, CV_DEFAULT, NULL);
+	mutex_init(&mmp->mmp_io_lock, NULL, MUTEX_DEFAULT, NULL);
+}
+
+void
+mmp_fini(dsl_pool_t *dp)
+{
+	mmp_thread_state_t *mmp = &dp->dp_mmp;
+
+	mutex_destroy(&mmp->mmp_thread_lock);
+	cv_destroy(&mmp->mmp_thread_cv);
+	mutex_destroy(&mmp->mmp_io_lock);
+	bzero(mmp, sizeof (mmp_thread_state_t));
+}
+
+static void
+mmp_thread_enter(mmp_thread_state_t *mmp)
+{
```

Comment:
We should use the `CALLB_CPR_INIT` function here, and `CALLB_CPR_EXIT` in `mmp_thread_exit()` since it's what they'll expect in upstream OpenZFS.  See `txg_thread_enter/txg_thread_exit`.